### PR TITLE
drivers: add support for velodyne HDL32E

### PR DESCRIPTION
**Pull request by fengqikai1414**

### CODE REVIEW

File: modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include <cmath>
+#include <string>
+#include <thread>
+
+namespace apollo {
+namespace drivers {
+namespace velodyne {
+
+Velodyne32Driver::Velodyne32Driver(const Config& config) { 
+  config_ = config;
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: It is suggested to use initialization list.

File: modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+namespace drivers {
+namespace velodyne {
+
+Velodyne32Driver::Velodyne32Driver(const Config& config) { 
+  config_ = config; 
+}
+
+void Velodyne32Driver::init(ros::NodeHandle& node) {
+  double packet_rate = 0.0;  // packet frequency (Hz)
+  packet_rate = 1808.0;
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: 35 and 36 should be done in one line. Also, please use const for number 1808.0.

File: modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  double frequency = (config_.rpm / 60.0);  // expected Hz rate
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: Define const for 60.0

File: modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ *  @returns true unless end of file reached
+ */
+bool Velodyne32Driver::poll(void) {
+  // Allocate a new shared pointer for zero-copy sharing with other nodelets.
+  velodyne_msgs::VelodyneScanUnifiedPtr scan(
+      new velodyne_msgs::VelodyneScanUnified);
+
+  if (basetime_ == 0) {
+    usleep(100);
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: usleep is deprecated. Use nanosleep or c++ version of sleep.

File: modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    return true;
+  }
+
+  // publish message using time of last packet read
+  ROS_DEBUG("Publishing a full Velodyne scan.");
+  scan->header.stamp = ros::Time().now();
+  scan->header.frame_id = config_.frame_id;
+  // we use fisrt packet gps time update gps base hour
+  // in cloud nodelet, will update base time packet by packet
+  uint32_t current_secs = *((uint32_t*)(&scan->packets.front().data[0] + 1200));
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: Define const for 1200.

Comment (lianglia-apollo) [reply modules/drivers/velodyne/velodyne_driver/src/driver/driver32.cpp]: Use reinterpret_cast for pointers.

File: modules/drivers/velodyne/velodyne_pointcloud/include/velodyne_pointcloud/velodyne_parser.h [hunk 6]
```diff
@@ -211,12 +211,8 @@ class VelodyneParser {
...
   virtual void order(VPointCloud::Ptr &cloud) = 0;
 
-  const Calibration &get_calibration() {
-    return calibration_;
-  }
-  const double get_last_timestamp() {
-    return last_time_stamp_;
-  }
+  const Calibration &get_calibration() { return calibration_; }
+  const double get_last_timestamp() { return last_time_stamp_; }
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_pointcloud/include/velodyne_pointcloud/velodyne_parser.h]: I think const double as return will have a compiling warning. Use double for return type.

File: modules/drivers/velodyne/velodyne_pointcloud/src/lib/velodyne32_parser.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  double timestamp = get_gps_stamp(t, previous_packet_stamp_, gps_base_usec_);
+  return timestamp;
+}
+
+void Velodyne32Parser::unpack(const velodyne_msgs::VelodynePacket& pkt,
+                              VPointCloud& pc) {
+  const RawPacket* raw = (const RawPacket*)&pkt.data[0];
+  double basetime = raw->gps_timestamp;  // usec
+
+  for (int i = 0; i < BLOCKS_PER_PACKET; i++) {  // 12
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_pointcloud/src/lib/velodyne32_parser.cpp]: ++i

File: modules/drivers/velodyne/velodyne_pointcloud/src/lib/velodyne32_parser.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+#include <ros/ros.h>
+
+namespace apollo {
+namespace drivers {
+namespace velodyne {
+
+Velodyne32Parser::Velodyne32Parser(Config config)
+    : VelodyneParser(config), gps_base_usec_(0), previous_packet_stamp_(0) {
+  inner_time_ = &velodyne::INNER_TIME_HDL32E;
+  need_two_pt_correction_ = false;
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_pointcloud/src/lib/velodyne32_parser.cpp]: Use initilization list dor innner_time_ and need_two_pt_correction_

File: modules/drivers/velodyne/velodyne_pointcloud/include/velodyne_pointcloud/velodyne_parser.h [hunk 9]
```diff
@@ -289,6 +285,26 @@ class Velodyne64Parser : public VelodyneParser {
 
 };  // class Velodyne64Parser
 
+class Velodyne32Parser : public VelodyneParser {
+ public:
+  Velodyne32Parser(Config config);
```
Comment (lianglia-apollo) [inline modules/drivers/velodyne/velodyne_pointcloud/include/velodyne_pointcloud/velodyne_parser.h]: use const reference for config

**Comment by quning78** [comment]
@fengqikai1414 in next commit, please fix the problems mentioned in Liangliang's comments

**Pull request merged**